intermediate_notebooks/E2E/taxi/NYCTaxi-E2E-GPU-Dask.ipynb

```diff
@@ -0,0 +1,1643 @@
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# reset_index and drop index column\n",
+    "taxi_df = taxi_df.reset_index(drop=True)\n",
+    "taxi_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Adding Interesting Features\n",
+    "\n",
+    "Dask & cuDF provide standard DataFrame operations, but also let you run \"user defined functions\" on the underlying data.\n",
+    "\n",
+    "cuDF's [apply_rows](https://rapidsai.github.io/projects/cudf/en/0.6.0/api.html#cudf.dataframe.DataFrame.apply_rows) operation is similar to Pandas's [DataFrame.apply](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html), except that for cuDF, custom Python code is [JIT compiled by numba](https://numba.pydata.org/numba-doc/dev/cuda/kernels.html) into GPU kernels.\n",
+    "\n",
+    "We'll use a Haversine Distance calculation to find total trip distance, and extract additional useful variables from the datetime fields."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## add features\n",
+    "\n",
+    "taxi_df['hour'] = taxi_df['pickup_datetime'].dt.hour\n",
+    "taxi_df['year'] = taxi_df['pickup_datetime'].dt.year\n",
+    "taxi_df['month'] = taxi_df['pickup_datetime'].dt.month\n",
+    "taxi_df['day'] = taxi_df['pickup_datetime'].dt.day\n",
+    "taxi_df['day_of_week'] = taxi_df['pickup_datetime'].dt.weekday\n",
+    "taxi_df['is_weekend'] = (taxi_df['day_of_week']>=5).astype(np.int32)\n",
+    "\n",
+    "#calculate the time difference between dropoff and pickup.\n",
+    "taxi_df['diff'] = taxi_df['dropoff_datetime'].astype('int64') - taxi_df['pickup_datetime'].astype('int64')\n",
+    "taxi_df['diff']=(taxi_df['diff']/1000).astype('int64')\n",
+    "\n",
+    "taxi_df['pickup_latitude_r'] = taxi_df['pickup_latitude']//.01*.01\n",
+    "taxi_df['pickup_longitude_r'] = taxi_df['pickup_longitude']//.01*.01\n",
+    "taxi_df['dropoff_latitude_r'] = taxi_df['dropoff_latitude']//.01*.01\n",
+    "taxi_df['dropoff_longitude_r'] = taxi_df['dropoff_longitude']//.01*.01\n",
+    "\n",
+    "taxi_df = taxi_df.drop('pickup_datetime', axis=1)\n",
+    "taxi_df = taxi_df.drop('dropoff_datetime', axis=1)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def haversine_dist(df):\n",
+    "    h_distance = cuspatial.haversine_distance(df['pickup_longitude'], df['pickup_latitude'], df['dropoff_longitude'], df['dropoff_latitude'])\n",
+    "    df['h_distance']= h_distance\n",
+    "    df['h_distance']= df['h_distance'].astype('float32')\n",
+    "    return df\n",
+    "\n",
+    "taxi_df = taxi_df.map_partitions(haversine_dist)\n",
+    "taxi_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Some EDA"
```

Comment:
Maybe include a markdown cell above explaining what you hope to accomplish with what I assume is "Exploratory Data Analysis"?

I'd have expected to see this section above the cell where we use `.query` to filter out rows with non-sensical records, to show that there are such records and that we need to filter them.